fv/bpf_test.go

```diff
@@ -831,6 +831,203 @@ func describeBPFTests(opts ...bpfTestOpt) bool {
+					tgtPort := 8055
+					externalIP := []string{"10.1.2.3"}
+
+					BeforeEach(func() {
+						srcIPRange := []string{}
+						testSvc = k8sLBService(testSvcName, "10.101.0.10", w[0][0], 80, tgtPort, testOpts.protocol, externalIP, srcIPRange)
+						testSvcNamespace = testSvc.ObjectMeta.Namespace
+						_, err := k8sClient.CoreV1().Services(testSvcNamespace).Create(testSvc)
+						Expect(err).NotTo(HaveOccurred())
+						Eventually(k8sGetEpsForServiceFunc(k8sClient, testSvc), "10s").Should(HaveLen(1),
+							"Service endpoints didn't get created? Is controller-manager happy?")
+						if testOpts.connTimeEnabled {
+							Skip("FIXME externalClient also does conntime balancing")
+						}
+						pol.Spec.Ingress = []api.Rule{
+							{
+								Action: "Allow",
+								Source: api.EntityRule{
+									Nets: []string{
+										externalClient.IP + "/32",
+									},
+								},
+							},
+						}
+						pol = updatePolicy(pol)
+
+					})
+					It("should have connectivity from external Client via a service to workload 0", func() {
+						ip := testSvc.Spec.ExternalIPs
+						port := uint16(testSvc.Spec.Ports[0].Port)
+						felixes[1].Exec("ip", "route", "add", "local", "10.1.2.3", "dev", "eth0")
+						felixes[0].Exec("ip", "route", "add", "local", "10.1.2.3", "dev", "eth0")
+						externalClient.Exec("ip", "route", "add", "10.1.2.3", "via", felixes[0].IP)
+						externalClient.EnsureBinary("test-connection")
+						cc.ExpectSome(externalClient, TargetIP(ip[0]), port)
+						cc.CheckConnectivity()
+					})
+				})
+
+				Context("Test LB-service with src ranges", func() {
+					var (
+						testSvc          *v1.Service
+						testSvcNamespace string
+					)
+					testSvcName := "test-lb-service-extip"
+					tgtPort := 8055
+					externalIP := []string{"10.1.2.3"}
+					srcIPRange := []string{"10.65.1.3/24"}
+
+					BeforeEach(func() {
+						testSvc = k8sLBService(testSvcName, "10.101.0.10", w[0][0], 80, tgtPort, testOpts.protocol, externalIP, srcIPRange)
+						testSvcNamespace = testSvc.ObjectMeta.Namespace
+						_, err := k8sClient.CoreV1().Services(testSvcNamespace).Create(testSvc)
+						Expect(err).NotTo(HaveOccurred())
+						Eventually(k8sGetEpsForServiceFunc(k8sClient, testSvc), "10s").Should(HaveLen(1),
+							"Service endpoints didn't get created? Is controller-manager happy?")
+
+					})
+					It("should have connectivity from workloads[1][0],[1][1] via a service to workload 0", func() {
+						ip := testSvc.Spec.ExternalIPs
+						port := uint16(testSvc.Spec.Ports[0].Port)
+						felixes[1].Exec("ip", "route", "add", "local", "10.1.2.3", "dev", "eth0")
+						felixes[0].Exec("ip", "route", "add", "local", "10.1.2.3", "dev", "eth0")
+						cc.ExpectSome(w[1][0], TargetIP(ip[0]), port)
+						cc.ExpectSome(w[1][1], TargetIP(ip[0]), port)
+						cc.ExpectNone(w[0][1], TargetIP(ip[0]), port)
+						cc.CheckConnectivity()
+					})
+				})
+
+				Context("Test LB-service with external Client, LB src ranges", func() {
```

Comment:
This test and the one below act as a pair to verify that the external client is only allowed if its IP is in the range.  IMO, it's worth engineering the two tests to share code so that all the set-up is identical apart from the source IP and the expectation of whether the connection is allowed.  One way to do that is to make a table-driven test where each entry in the table is a pair or source IP range and expected outcome.  Another would be to move all the common set-up to a shared before each so that the BeforeEach for each test is only tweaking the source IPs.

Otherwise, a bug in the copying of the set-up here could make the test pass/fail for the wrong reason.